x/emissions/types/events.go

```diff
@@ -0,0 +1,462 @@
+package types
+
+import (
+	"encoding/json"
+	fmt "fmt"
+
+	alloraMath "github.com/allora-network/allora-chain/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+const (
+	EventTypeInfererScoresSet              = "inferer_scores_set"
+	EventTypeForecasterScoresSet           = "forecaster_scores_set"
+	EventTypeReputerScoresSet              = "reputer_scores_set"
+	EventTypeInfererRewardsSet             = "inferer_rewards_set"
+	EventTypeForecasterRewardsSet          = "forecaster_rewards_set"
+	EventTypeReputerAndDelegatorRewardsSet = "reputer_and_delegator_rewards_set"
+
+	AttributeKeyTopicId     = "topic_id"
+	AttributeKeyBlockHeight = "block_height"
+	AttributeKeyAddres      = "address"
+	AttributeKeyScores      = "scores"
+	AttributeKeyRewards     = "rewards"
+)
+
+/// Emitters
+
+func EmitNewInfererScoresSetEvent(ctx sdk.Context, scores []Score) {
+	ctx.EventManager().EmitEvent(NewScoresSetEventBase(EventTypeInfererRewardsSet, scores))
+}
+
+func EmitNewForecasterScoresSetEvent(ctx sdk.Context, scores []Score) {
+	ctx.EventManager().EmitEvent(NewScoresSetEventBase(EventTypeForecasterScoresSet, scores))
+}
+
+func EmitNewReputerScoresSetEvent(ctx sdk.Context, scores []Score) {
+	ctx.EventManager().EmitEvent(NewScoresSetEventBase(EventTypeReputerScoresSet, scores))
+}
+
+func EmitNewInfererRewardsSetEvent(ctx sdk.Context, blockHeight BlockHeight, rewards []TaskReward) {
+	ctx.EventManager().EmitEvent(NewRewardsSetEventBase(EventTypeInfererRewardsSet, blockHeight, rewards))
+}
+
+func EmitNewForecasterRewardsSetEvent(ctx sdk.Context, blockHeight BlockHeight, rewards []TaskReward) {
+	ctx.EventManager().EmitEvent(NewRewardsSetEventBase(EventTypeForecasterRewardsSet, blockHeight, rewards))
+}
+
+func EmitNewReputererAndDelegaterRewardsSetEvent(ctx sdk.Context, blockHeight BlockHeight, rewards []TaskReward) {
+	ctx.EventManager().EmitEvent(NewRewardsSetEventBase(EventTypeReputerAndDelegatorRewardsSet, blockHeight, rewards))
+}
+
+/// Utils
+
+func NewScoresSetEventBase(eventType string, scores []Score) sdk.Event {
+	if len(scores) == 0 {
+		return sdk.Event{}
+	}
+	topicId := scores[0].TopicId
+	blockHeight := scores[0].BlockHeight
+	addresses := make([]string, len(scores))
+	scoreValues := make([]alloraMath.Dec, len(scores))
+	for i, score := range scores {
+		addresses[i] = score.Address
+		scoreValues[i] = score.Score
+	}
+	return sdk.NewEvent(
+		eventType,
+		sdk.NewAttribute(AttributeKeyTopicId, FormatNumberForEventAttribute(topicId)),
+		sdk.NewAttribute(AttributeKeyTopicId, FormatNumberForEventAttribute(blockHeight)),
+		sdk.NewAttribute(AttributeKeyTopicId, FormatStringsForEventAttribute(addresses)),
+		sdk.NewAttribute(AttributeKeyScores, FormatDecsForEventAttribute(scoreValues)),
+	)
+}
+
+func NewRewardsSetEventBase(eventType string, blockHeight BlockHeight, rewards []TaskReward) sdk.Event {
+	if len(rewards) == 0 {
+		return sdk.Event{}
+	}
```

Comment:
is this what you want? To still emit an event? Maybe return nil and have the outer function handle that and not emit?

Reply:
no I don't...